File: app/src/main/java/com/stedi/randomimagegenerator/app/presenter/impl/ColoredNoiseParamsPresenterImpl.java
package com.stedi.randomimagegenerator.app.presenter.impl;

import android.support.annotation.NonNull;

import com.stedi.randomimagegenerator.app.model.data.PendingPreset;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.ColoredNoiseParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.base.EffectGeneratorParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.base.GeneratorParams;
import com.stedi.randomimagegenerator.app.presenter.interfaces.ColoredNoiseParamsPresenter;
import com.stedi.randomimagegenerator.generators.ColoredNoiseGenerator;

public class ColoredNoiseParamsPresenterImpl implements ColoredNoiseParamsPresenter {
    private final PendingPreset pendingPreset;

    private UIImpl ui;
    private ColoredNoiseParams params;

    public ColoredNoiseParamsPresenterImpl(@NonNull PendingPreset pendingPreset) {
        this.pendingPreset = pendingPreset;
    }

    @Override
    public void getValues() {
        ui.showOrientation(params.getNoiseOrientation());
        ui.showType(params.getNoiseType());
    }

    @Override
    public void setOrientation(@NonNull ColoredNoiseGenerator.Orientation orientation) {
        params.setNoiseOrientation(orientation);
    }

    @Override
    public void setType(@NonNull ColoredNoiseGenerator.Type type) {
        params.setNoiseType(type);
    }

    @Override
    public void onAttach(@NonNull UIImpl ui) {
        this.ui = ui;
        GeneratorParams currentParams = pendingPreset.getCandidate().getGeneratorParams();
        if (currentParams instanceof EffectGeneratorParams) {
            params = (ColoredNoiseParams) ((EffectGeneratorParams) currentParams).getTarget();
        } else {
            params = (ColoredNoiseParams) currentParams;
        }
    }

    @Override
    public void onDetach() {
        this.ui = null;
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/presenter/impl/ColoredNoiseParamsPresenterImpl.kt
package com.stedi.randomimagegenerator.app.presenter.impl

import com.stedi.randomimagegenerator.app.model.data.PendingPreset
import com.stedi.randomimagegenerator.app.model.data.Preset
import com.stedi.randomimagegenerator.app.model.data.generatorparams.ColoredNoiseParams
import com.stedi.randomimagegenerator.app.model.data.generatorparams.base.EffectGeneratorParams
import com.stedi.randomimagegenerator.app.presenter.interfaces.ColoredNoiseParamsPresenter
import com.stedi.randomimagegenerator.app.presenter.interfaces.SimpleIntegerParamsPresenter
import com.stedi.randomimagegenerator.generators.ColoredNoiseGenerator
import javax.inject.Inject

class ColoredNoiseParamsPresenterImpl @Inject constructor(
        private val pendingPreset: PendingPreset) : SimpleIntegerParamsPresenterImpl(pendingPreset), ColoredNoiseParamsPresenter {

    private val candidate: Preset
        get() = pendingPreset.getCandidate()
                ?: throw IllegalStateException("pending preset candidate must not be null")

    private var ui: ColoredNoiseParamsPresenter.UIImpl? = null

    private lateinit var params: ColoredNoiseParams

    override fun onAttach(ui: SimpleIntegerParamsPresenter.UIImpl) {
        super.onAttach(ui)
        this.ui = ui as ColoredNoiseParamsPresenter.UIImpl
        val currentParams = candidate.getGeneratorParams()
        if (currentParams is EffectGeneratorParams) {
            params = currentParams.target as ColoredNoiseParams
        } else {
            params = currentParams as ColoredNoiseParams
        }
    }

    override fun onDetach() {
        super.onDetach()
        this.ui = null
    }

    override fun getValues() {
        super.getValues()
        ui?.showOrientation(params.noiseOrientation)
        ui?.showType(params.noiseType)
    }

    override fun setOrientation(orientation: ColoredNoiseGenerator.Orientation) {
        params.noiseOrientation = orientation
    }

    override fun setType(type: ColoredNoiseGenerator.Type) {
        params.noiseType = type
    }
}